[Sema] Implement support for -Wformat-signedness

karka228 opened the pull request:
In gcc there exist a modifier option -Wformat-signedness that turns on additional signedness warnings in the already existing -Wformat warning.

This patch implements that support in clang.

karka228 committed:
[Sema] Implement support for -Wformat-signedness

In gcc there exist a modifier option -Wformat-signedness that turns on
additional signedness warnings in the already existing -Wformat warning.

This patch implements that support in clang.

karka228 committed:
[Sema] Implement support for -Wformat-signedness

Move the testcase to its correct dir.

hazohelet commented on clang/include/clang/Driver/Options.td:
```diff
@@ -918,6 +918,9 @@ def Wdeprecated : Flag<["-"], "Wdeprecated">, Group<W_Group>,
   HelpText<"Enable warnings for deprecated constructs and define __DEPRECATED">;
 def Wno_deprecated : Flag<["-"], "Wno-deprecated">, Group<W_Group>,
   Visibility<[ClangOption, CC1Option]>;
+def Wformat_signedness : Flag<["-"], "Wformat-signedness">,
```
I think the warning option config should go to `clang/include/clang/Basic/DiagnosticSemaKinds.td` instead of the toplevel driver setting. For example,  you wouldn't have `-Wno-format-signedness` atm.

karka228 replied:
I have tried to mimic the gcc behavior when implementing this and in gcc it seems like -Wformat-signedness don't introduce a new warning but only modify the behavior of the already existing -Wformat warning.
If I place the warning option config in clang/include/clang/Basic/DiagnosticSemaKinds.td as you suggest don't I then also create a new warning with a new warning text? ... that is not what gcc do (as far as I understand).

Well, with that said we don't have to follow exactly how gcc implemented this if we think there is a better way to implement this in clang.

I agree that support for -Wno-format-signednes is missing from my patch and that should be added.

apple-fcloutier replied:
There's a lot of good, free integration that you get with a real warning flag:

* -Wno symmetry (as @hazohelet already said)
* control via `#pragma clang diagnostic`
* diagnostic text shows the name of the warning flag (in case you need to look it up or disable it)

-Wformat-pedantic is another warning that "modifies" -Wformat and it exists as its own warning. I think that it would make sense to do the same thing with -Wformat-signedness.

inclyc replied:
I agree with @hazohelet

AaronBallman replied:
Yup, this will have to change to adding the diagnostic to DiagnosticSemaKinds.td. I think we will want to control this separately from `-Wformat` but have it grouped under that flag.

apple-fcloutier commented on clang/test/Sema/format-strings-signedness.c:
```diff
@@ -0,0 +1,107 @@
+// RUN: %clang_cc1 -std=c11 -fsyntax-only -verify -Wformat -Wformat-signedness %s
+
+int printf(const char *restrict format, ...);
+int scanf(const char * restrict, ...);
```
Please add tests with enum values:

```c
#include <limits.h>

enum foo {
	minus_one = -1,
	int_val = INT_MAX,
	unsigned_val = (unsigned)INT_MIN
};
```

inclyc commented on clang/test/Sema/format-strings-signedness.c:
```diff
@@ -0,0 +1,107 @@
+{
+    printf("%d", x); // no-warning
+    printf("%u", x); // expected-warning{{format specifies type 'unsigned int' but the argument has type 'int'}}
+    printf("%x", x); // expected-warning{{format specifies type 'unsigned int' but the argument has type 'int'}}
```
will this be a little pedantic?

inclyc commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -11815,6 +11815,18 @@ isArithmeticArgumentPromotion(Sema &S, const ImplicitCastExpr *ICE) {
          S.Context.getFloatingTypeOrder(From, To) < 0;
 }
 
+static clang::analyze_format_string::ArgType::MatchKind
```
```suggestion
static analyze_format_string::ArgType::MatchKind
```

Maybe we can remove this namespace?

inclyc reviewed (commented):
Also please add release notes to let users know what has been changed :)

See: `clang/docs/ReleaseNotes.rst`

AaronBallman commented on clang/include/clang/AST/FormatString.h:
```diff
@@ -264,6 +264,8 @@ class ArgType {
     /// The conversion specifier and the argument type are compatible. For
     /// instance, "%d" and int.
     Match = 1,
+    /// The conversion specifier and the argument type have different sign
```
```suggestion
    /// The conversion specifier and the argument type have different sign.
```

AaronBallman reviewed (commented):
I'd like to understand the need for this diagnostic a bit more. I realize GCC has it, but it's off-by-default in GCC and `-Wformat` does not enable it either: https://godbolt.org/z/Gxczsjdj5

We believe that users don't enable off by default warnings often enough to warrant supporting them, and this feels like a very pedantic warning in many cases. e.g., `printf("%u", 0);`

karka228 commented:
I have started to implement defining the -Wformat-signedness option config in clang/include/clang/Basic/DiagnosticSemaKinds.td as suggested originally by @hazohelet and I agree that the implementation is a lot cleaner that way.
However before finishing implementing that I think we need to conclude what level of gcc compatibility that we aim for in clang.

Below I list the characteristics of -Wformat-signedness that I have seen by testing it out in gcc and reading the gcc documentation (I have not inspected the gcc source code):

1. The option -Wformat-signedness is default off.
2. The -Wformat-signedness warnings are not enabled alone by the -Wformat option.
3. The -Wformat-signedness warnings are not enabled alone by the -Wformat-signedness option.
4. The -Wformat-signedness warnings are enabled by the option -Wformat together with the option -Wformat-signedness.
5. Parts of the -Wformat-signedness warnings (regarding scanf) are enabled by the options -Wformat together with the option -pedantic.
6. Warnings produced by -Wformat-signedness is reported as a -Wformat warnings (e.g "warning: format ‘%u’ expects argument of type ‘unsigned int’, but argument 2 has type ‘int’ [-Wformat=]"
7. Warnings produced by -Wformat-signedness can be suppressed by "#pragma GCC diagnostic ignored -Wformat"

Which of the above points are important for us to follow in clang to get a reasonable gcc compatibility?
To me I think point 1, 2 and maybe 7 is the most important to follow.

karka228 committed:
[Sema] Implement support for -Wformat-signedness

Move warning option config to DiagnosticSemaKinds.td

karka228 committed:
[Sema] Implement support for -Wformat-signedness

Reformat code with clang-format.

karka228 committed:
[Sema] Implement support for -Wformat-signedness

Fixed and extended the enum tests in format-strings-signedness.c and
format-strings-signedness-fixit.c

AaronBallman commented:
> I have started to implement defining the -Wformat-signedness option config in clang/include/clang/Basic/DiagnosticSemaKinds.td as suggested originally by @hazohelet and I agree that the implementation is a lot cleaner that way. However before finishing implementing that I think we need to conclude what level of gcc compatibility that we aim for in clang.
> 
> Below I list the characteristics of -Wformat-signedness that I have seen by testing it out in gcc and reading the gcc documentation (I have not inspected the gcc source code):
> 
>     1. The option -Wformat-signedness is default off.
> 
>     2. The -Wformat-signedness warnings are not enabled alone by the -Wformat option.
> 
>     3. The -Wformat-signedness warnings are not enabled alone by the -Wformat-signedness option.
> 
>     4. The -Wformat-signedness warnings are enabled by the option -Wformat together with the option -Wformat-signedness.
> 
>     5. Parts of the -Wformat-signedness warnings (regarding scanf) are enabled by the options -Wformat together with the option -pedantic.
> 
>     6. Warnings produced by -Wformat-signedness is reported as a -Wformat warnings (e.g "warning: format ‘%u’ expects argument of type ‘unsigned int’, but argument 2 has type ‘int’ [-Wformat=]"
> 
>     7. Warnings produced by -Wformat-signedness can be suppressed by "#pragma GCC diagnostic ignored -Wformat"
> 
> 
> Which of the above points are important for us to follow in clang to get a reasonable gcc compatibility? To me I think point 1, 2 and maybe 7 is the most important to follow.

I agree, 1, 2, 4 (same idea as 2, basically), and 7 are the critical things we'd need for compatibility. I personally find 3, 5, and 6 to be kind of strange behavior  though I can see how it would arise.

karka228 commented:
> > 1. The option -Wformat-signedness is default off.
> > 
> > 2. The -Wformat-signedness warnings are not enabled alone by the -Wformat option.
> > 
> > 3. The -Wformat-signedness warnings are not enabled alone by the -Wformat-signedness option.
> > 
> > 4. The -Wformat-signedness warnings are enabled by the option -Wformat together with the option -Wformat-signedness.
> > 
> > 5. Parts of the -Wformat-signedness warnings (regarding scanf) are enabled by the options -Wformat together with the option -pedantic.
> > 
> > 6. Warnings produced by -Wformat-signedness is reported as a -Wformat warnings (e.g "warning: format ‘%u’ expects argument of type ‘unsigned int’, but argument 2 has type ‘int’ [-Wformat=]"
> > 
> > 7. Warnings produced by -Wformat-signedness can be suppressed by "#pragma GCC diagnostic ignored -Wformat"
>
> I agree, 1, 2, 4 (same idea as 2, basically), and 7 are the critical things we'd need for compatibility. I personally find 3, 5, and 6 to be kind of strange behavior though I can see how it would arise.

Good, then I know what the implementation have to adhere too. I agree that there are some strange behavior in the gcc implementation.

I think the first commit in my warn_format_signedness branch adhered to the important 1, 2, 4 and 7 points, but it implemented -Wformat-signedness as a simple clang option.

However when defining the option config in DiagnosticSemaKinds.td it will be a bit harder to implement. Too be able to handle 7) I simply have to add -Wformat-signedness to the -Wformat group, but then the implementation will violate 2). As far as I understand this can't be configured with the current framework. I need to extend the tablegen backend ClangDiagnosticsEmitter with some kind of new option to handle this, right?

hazohelet commented:
> I need to extend the tablegen backend ClangDiagnosticsEmitter with some kind of new option to handle this, right?

Alternatively, you could probably use `DiagnosticsEngine::isIgnored` to check if the `-Wformat-signedness` is enabled or not, and control whether `MatchSignedNess` is `NoMatch` or `Match`, as your first implementation did.
This way we can achieve GCC compatibility for 1, 2, 3, 4, 6, 7, while benefitting from a real warning flag.
However, in this implementation `-Wformat-signedness` would be an _dummy warning_ which never gets emitted. As far as I know, there isn't anything similar in the codebase, so it might not be the ideal way.

karka228 commented:
> Alternatively, you could probably use `DiagnosticsEngine::isIgnored` to check if the `-Wformat-signedness` is enabled or not, and control whether `MatchSignedNess` is `NoMatch` or `Match`, as your first implementation did. This way we can achieve GCC compatibility for 1, 2, 3, 4, 6, 7, while benefitting from a real warning flag. However, in this implementation `-Wformat-signedness` would be an _dummy warning_ which never gets emitted. As far as I know, there isn't anything similar in the codebase, so it might not be the ideal way.

A bit ugly with a _dummy warning_, but it simplifies the implementation. What do you other think? Is this an acceptable solution or not?

karka228 committed:
[Sema] Implement support for -Wformat-signedness

Add more GCC compatibility.

karka228 committed:
[Sema] Implement support for -Wformat-signedness

Fixed bug and extended the testcase to cover this.

karka228 commented:
Ping.

AaronBallman commented:
> > I need to extend the tablegen backend ClangDiagnosticsEmitter with some kind of new option to handle this, right?
> 
> Alternatively, you could probably use `DiagnosticsEngine::isIgnored` to check if the `-Wformat-signedness` is enabled or not, and control whether `MatchSignedNess` is `NoMatch` or `Match`, as your first implementation did. This way we can achieve GCC compatibility for 1, 2, 3, 4, 6, 7, while benefitting from a real warning flag. However, in this implementation `-Wformat-signedness` would be an _dummy warning_ which never gets emitted. As far as I know, there isn't anything similar in the codebase, so it might not be the ideal way.

Sorry for the delayed response, this fell off my radar (thanks @Endilll for mentioning this after meeting with @hazohelet!).

This would be novel but might not be a terrible approach. I'm not convinced that #3 is a particularly good design (and #6 doesn't seem critical to me either), but I think modifying tablegen to improve support here would be a heavy lift for little benefit. I'd say let's try that approach and see how we like it, and go from there. WDYT?

karka228 commented:
> Sorry for the delayed response, this fell off my radar (thanks @Endilll for mentioning this after meeting with @hazohelet!).

No problem. As I wrote above I work in an out of tree backend and I have already implemented this in our downstream repo. Our customer that requested this already have this feature. To me this is not at all urgent.

> This would be novel but might not be a terrible approach. I'm not convinced that #3 is a particularly good design (and #6 doesn't seem critical to me either), but I think modifying tablegen to improve support here would be a heavy lift for little benefit. I'd say let's try that approach and see how we like it, and go from there. WDYT?

Sounds good. I think what is currently implemented in this pull request use @hazohelet  idea with the dummy warning. As this was started in December I think I have to rebase this onto a more recent main branch.

I will look through the code in the pull request (as it was a while ago since I looked at this) and do the rebase ...

AaronBallman commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -12465,6 +12465,20 @@ isArithmeticArgumentPromotion(Sema &S, const ImplicitCastExpr *ICE) {
+handleFormatSignedness(analyze_format_string::ArgType::MatchKind Match,
+                       DiagnosticsEngine &Diags, SourceLocation Loc) {
+  if (Match == analyze_format_string::ArgType::NoMatchSignedness) {
+    if (!Diags.isIgnored(
+            diag::warn_format_conversion_argument_type_mismatch_signedness,
+            Loc))
+      Match = analyze_format_string::ArgType::NoMatch;
+    else
+      Match = analyze_format_string::ArgType::Match;
```
```suggestion
   Match = Diags.isIgnored(
            diag::warn_format_conversion_argument_type_mismatch_signedness,
            Loc) ? analyze_format_string::ArgType::Match : analyze_format_string::ArgType::NoMatch;
```
Probably need to be reformatted, but makes the code a little bit more clear (IMO).

AaronBallman reviewed (commented):
Can you also add a test showing the difference between `-Wformat -Wformat-signedness` and `-Wformat-signedness` by itself (which does nothing)? I'd also like to see a test demonstrating that `#pragma GCC diagnostic ignored -Wformat` disables the signedness warnings.

Otherwise, this generally LGTM!

karka228 committed:
[Sema] Implement support for -Wformat-signedness

Misc fixes.

karka228 commented:
> Can you also add a test showing the difference between `-Wformat -Wformat-signedness` and `-Wformat-signedness` by itself (which does nothing)?

With the current implementation `-Wformat-signedness` by itself actually turn on the signedness warnings. This is not compatible with how gcc do it. I guess I have thought that it was not that important to be compatible with gcc in this respect. However if it is I guess this could be implemented.

> I'd also like to see a test demonstrating that `#pragma GCC diagnostic ignored -Wformat` disables the signedness warnings.

There is a single test in the end of clang/test/Sema/format-strings-signedness.c that demonstrate this.

AaronBallman commented:
> > Can you also add a test showing the difference between `-Wformat -Wformat-signedness` and `-Wformat-signedness` by itself (which does nothing)?
> 
> With the current implementation `-Wformat-signedness` by itself actually turn on the signedness warnings. This is not compatible with how gcc do it. I guess I have thought that it was not that important to be compatible with gcc in this respect. However if it is I guess this could be implemented.

Personally, I find GCC's behavior a bit surprising, so I'm okay with a divergence there. But we should have tests for this just the same:

* `-Wformat-signedness`
* `-Wformat-signedness -Wno-format`
* `-Wno-format -Wformat-signedness`

The first test helps demonstrate that we intentionally deviate from GCC, but it also helps show whether `-Wformat-signedness` implies `-Wformat` or not. The second test helps answer whether `-Wformat-signedness` is turned off by disabling `-Wformat`. The third test helps answer whether format signedness can be turned on while turning off all other formatting warnings -- this test may not be needed if `-Wformat-signedness` by itself enables only the signedness diagnostics.

> > I'd also like to see a test demonstrating that `#pragma GCC diagnostic ignored -Wformat` disables the signedness warnings.
> 
> There is a single test in the end of clang/test/Sema/format-strings-signedness.c that demonstrate this.

Ah thank you!

karka228 committed:
[Sema] Implement support for -Wformat-signedness

Updated testcase.

karka228 commented:
> > With the current implementation `-Wformat-signedness` by itself actually turn on the signedness warnings. This is not compatible with how gcc do it. I guess I have thought that it was not that important to be compatible with gcc in this respect. However if it is I guess this could be implemented.
> 
> Personally, I find GCC's behavior a bit surprising, so I'm okay with a divergence there.

Sorry, I'm confused. Ignore what I said above. What is not compatible with gcc is only that -Wformat is turned on default by clang and not by gcc. The handling of `-Wformat-signedness` is as far as I can tell compatible with gcc.

> But we should have tests for this just the same:
> 
>     * `-Wformat-signedness`
> 
>     * `-Wformat-signedness -Wno-format`
> 
>     * `-Wno-format -Wformat-signedness`

I agree, it is good to have tests of those combinations. I have now added them to clang/test/Sema/format-strings-signedness.c

AaronBallman commented on clang/test/Sema/format-strings-signedness.c:
```diff
@@ -0,0 +1,223 @@
+// RUN: %clang_cc1 -triple=x86_64-pc-linux-gnu -std=c11 -fsyntax-only -verify -Wformat-signedness %s
+// RUN: %clang_cc1 -triple=x86_64-pc-win32 -std=c11 -fsyntax-only -verify -Wformat-signedness %s
+
+// Verify that -Wformat-signedness warnings are not reported with only -Wformat
+// (gcc compat).
+// RUN: %clang_cc1 -triple=x86_64-pc-linux-gnu -std=c11 -fsyntax-only -Wformat -Werror %s
+
+// Verify that -Wformat-signedness with -Wno-format are not reported (gcc compat).
+// RUN: %clang_cc1 -triple=x86_64-pc-linux-gnu -std=c11 -fsyntax-only -Wformat-signedness -Wno-format -Werror %s
+// RUN: %clang_cc1 -triple=x86_64-pc-linux-gnu -std=c11 -fsyntax-only -Wno-format -Wformat-signedness -Werror %s
```
```suggestion
// Verify that -Wformat-signedness warnings are not reported with only -Wformat
// (gcc compat).
// RUN: %clang_cc1 -triple=x86_64-pc-linux-gnu -std=c11 -fsyntax-only -Wformat -Werror %s

// Verify that -Wformat-signedness with -Wno-format are not reported (gcc compat).
// RUN: %clang_cc1 -triple=x86_64-pc-linux-gnu -std=c11 -fsyntax-only -Wformat-signedness -Wno-format -verify=okay %s
// RUN: %clang_cc1 -triple=x86_64-pc-linux-gnu -std=c11 -fsyntax-only -Wno-format -Wformat-signedness -verify=okay %s

// okay-no-diagnostics
```
These are all missing `-verify` lines so they're relying on `-Werror` to upgrade diagnostics from warnings to errors and that the return value from `clang -cc1` will be non-zero if a diagnostic was emitted. This works but is a bit subtle -- usually we'd have `-verify` in all of the `RUN` lines and we'd use a different prefix for these and we ensure there are no diagnostics with the diagnostics verifier, along these lines.

karka228 replied:
Ohh, I didn't know that you could supply a prefix to `-verify`.  I will fix this.

karka228 committed:
[Sema] Implement support for -Wformat-signedness

Minor fixes of testcase

karka228 commented on clang/test/Sema/format-strings-signedness.c:
```diff
@@ -0,0 +1,223 @@
+// RUN: %clang_cc1 -triple=x86_64-pc-linux-gnu -std=c11 -fsyntax-only -Wformat-signedness -Wno-format -Werror %s
+// RUN: %clang_cc1 -triple=x86_64-pc-linux-gnu -std=c11 -fsyntax-only -Wno-format -Wformat-signedness -Werror %s
+
+#include <limits.h>
```
Sure I remove the include and use the expanded macro instead.

AaronBallman reviewed (approved):
LGTM!

karka228 commented:
> LGTM!

Thanks. I will rebase and then "squash and merge" this to the main branch.

jroelofs commented:
I think this diagnostic should not be firing on `%x` format specifiers:

https://clang.godbolt.org/z/YGh1qdq1h

```
#include <stdio.h>

void bad() {
    int u = 0;
    printf("%x\n", u);
}
```

```
<source>:5:20: warning: format specifies type 'unsigned int' but the argument has type 'int' [-Wformat]
    5 |     printf("%x\n", u);
      |             ~~     ^
      |             %x
1 warning generated.
```

karka228 commented:
> I think this diagnostic should not be firing on `%x` format specifiers:

In the C-standard (7.21.6.1 The fprintf function) when reading about the o,u,x,X format specifiers it say "The unsigned int argument is converted to unsigned octal (o), unsigned decimal (u), or unsigned hexadecimal notation (x or X) in the style dddd; the letters abcdef are used for x conversion and the letters ABCDEF for X conversion.". That's the reason why this diagnostic give a warning in this case. When this was implemented in clang we also said that we should follow the gcc behavior (as much as possible) as this warning have been available in gcc for a long time and gcc give a warning in this case.

jroelofs commented:
At the very least, the fixit shouldn't be recommending precisely what the user wrote, which funnily enough, gcc also gets wrong: https://clang.godbolt.org/z/88sqM5xET
